backend/src/api/song_routes.py

```diff
@@ -0,0 +1,179 @@
+from typing import TypedDict, NotRequired
+from flask import Blueprint, request
+from flask_login import login_required, current_user  # pyright: ignore
+from ..models import Song, likes_join, db
+from ..backend_api import GetSongs, ApiErrorResponse, IdAndTimestamps, GetSong, DeleteSong
+from ..forms.song_form import SongForm
+from datetime import datetime, timezone
+
+
+class GetSongWithNumLikes(TypedDict):
+    song_data: GetSong
+    num_likes: NotRequired[int]
+
+
+song_routes = Blueprint("songs", __name__)
+
+song_not_found_error: ApiErrorResponse = (
+    {"message": "Song Not Found", "errors": {"song_not_found_error": "This song could not be found"}},
+    404,
+)
+not_authorized_error: ApiErrorResponse = (
+    {
+        "message": "Not Authorized",
+        "errors": {"user_not_authorized_error": "You are not authorized to modify or delete this song"},
+    },
+    401,
+)
+
+
+def db_song_to_api_song(song: Song) -> GetSong:
+    api_song: GetSong = {
+        "id": song.id,
+        "name": song.name,
+        "artist_id": song.artist_id,
+        "song_ref": song.song_ref,
+        "created_at": str(song.created_at),
+        "updated_at": str(song.updated_at),
+    }
+
+    if song.thumb_url is not None:
+        api_song["thumb_url"] = song.thumb_url
+
+    if song.genre is not None:
+        api_song["genre"] = song.genre
+
+    return api_song
+
+
+# # I want a landing page of all songs (showing newest first)
+# # I want to be able to filter these results to show only songs uploaded by a specific artist, including possibly the current_user
+@song_routes.get("")
+def get_all_songs() -> GetSongs:
+    """
+    Check for query params first to see if we need to filter by an artist_id.
+    Query for all songs and return them in a list of song dictionaries.
+    """
+    artist_id: str | None = request.args.get("artist_id")
+
+    if artist_id and artist_id.isdigit():
+        id: int = int(artist_id)
+        songs = db.session.query(Song).filter(Song.artist_id == id).order_by(Song.created_at.desc())
+        return {"songs": [db_song_to_api_song(song) for song in songs]}
+
+    else:
+        songs = db.session.query(Song).order_by(Song.created_at.desc()).all()
+        return {"songs": [db_song_to_api_song(song) for song in songs]}
+
+
+# # I want to view a song's total likes
+@song_routes.get("/<int:song_id>")
+def get_song(
+    song_id: str,
```

Comment:
not blocking, but something joec noticed and that I later noticed reading his code, flask does actually convert `song_id` to an int here despite the lecture or reading that stated otherwise, it was just a wrong lecture, so we can accept an int type here, not a major deal since we do immediately convert to int, but worth noting

Reply:
I also like that there's no need for extra variable that is the int wrapped song_id.